### PR TITLE
util: add generic slice utilities Filter, Map, MapBy

### PR DESCRIPTION
**Pull request by chrisseto**
**util: add generic slice utilities Filter, Map, MapBy**
Devs will often find themselves repeating the same code within go due to the
lack of functional helpers like filter, map, and reduce. This can lead
to mistakes in the form of typos or overly terse naming, especially when
these operations need to be performed multiple times in a row.

This commit adds a minimal set of productivity helpers that are notably
missing from go's `exp/slices` packages.

In performance critical code, it may be undesirable to utilize generic
functions. It is also possible to overuse functional patterns and write
less understandable code. Devs will be ultimately responsible for
deciding when to wield these functional helpers.

Epic: none
Release note: None

### CODE REVIEW

**Comment by ecwall** [comment]
Have you looked at adding a 3rd party lib with a more complete API?

I think these would compose better if they are implemented as a monad like Stream in Java.

I like functional programming, but it seems to go against go's goal of maximizing boilerplate 😀.

**Comment by chrisseto** [comment]
> Have you looked at adding a 3rd party lib with a more complete API?

I intentionally decided to not integrate a 3rd party library. I don't want to change the entire style of the codebase, just add a few helpers to minimize friction when performing tedious and repetitive operations. I worry that an external library would lead to too much bike shedding as it'd be difficult to find the balance between too much and too little.

File: pkg/util/slices.go [hunk 1]
```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
 	}
 	return a
 }
+
+// Filter removes all elements from collection that do not satisfy predicate.
+// NOTE: Collection is modified in place. To create a new slice, utilize
```
Comment (rickystewart) [inline pkg/util/slices.go:56]: Drive-by: I would NEVER expect a `Filter()` function to modify a collection in-place. To me, this is asking for bugs. Up to you whether you want this behavior specifically for some reason (performance?)

Comment (chrisseto) [reply pkg/util/slices.go:56]: Yeah, I was thinking that it could be used in performance critical cases if we did the filter in place. I started to use this function elsewhere and I 100% agree with you.

I'll update it as soon as my working tree is clean 👍

Comment (rickystewart) [reply pkg/util/slices.go:56]: (Something to consider: a function name like `FilterInPlace()` would make clear the intent and avoid the possibility of bugs.)

Comment (chrisseto) [reply pkg/util/slices.go:56]: For the in place case, I think we'd want to fallback to utilizing a for loop? Or we can add `FilterInPlace` once it makes sense to do so :D

Comment (rickystewart) [reply pkg/util/slices.go:56]: I agree that if you want the in-place behavior (due to performance probably), you likely want to be writing a for-loop, taking advantage of function call inlining, etc.

Comment (chrisseto) [reply pkg/util/slices.go]: Fixed!

File: pkg/util/slices_test.go [hunk 2]
```diff
@@ -79,3 +79,67 @@ func TestCombinesUniqueStrings(t *testing.T) {
...
+
+func TestMapBy(t *testing.T) {
+	require.Equal(t, map[int]bool{}, MapBy(nil, func(i int) (int, bool) {
+		return 0, false
+	}))
+	require.Equal(t, map[int]bool{}, MapBy([]int{}, func(i int) (int, bool) {
+		return 0, false
+	}))
+	require.Equal(t, map[int]struct{}{
+		1: {},
```
Comment (rickystewart) [inline pkg/util/slices_test.go:137]: Nit: using some type besides `struct{}` would give a little more confidence that the function does what the documentation says. There's no distinguishing one `struct{}` from another. If you were using `int`'s for example, then the `require.Equal()` would be a stronger guarantee things are working.

File: pkg/util/slices_test.go [hunk 3]
```diff
@@ -79,3 +79,67 @@ func TestCombinesUniqueStrings(t *testing.T) {
...
+		return true
+	}))
+	require.Equal(t, []bool{false, true, false, true, false}, Map([]int{1, 2, 3, 4, 5}, func(i int) bool {
+		return i%2 == 0
+	}))
+}
+
+func TestMapBy(t *testing.T) {
+	require.Equal(t, map[int]bool{}, MapBy(nil, func(i int) (int, bool) {
+		return 0, false
```
Comment (rickystewart) [inline pkg/util/slices_test.go:131]: Nit: adding a `t.Fail()` in this function would help guarantee that this function is not called if the collection is empty, an important invariant.

File: pkg/util/slices.go [hunk 4]
```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
...
+//	// Construct a set.
+//	MapBy(numbers, func(i int) (int, struct{}) {
+//		return i, struct{}{}
+//	})
+//
+//	// Construct a map of numbers to their square.
+//	MapBy(numbers, func(i int) (int, int) {
+//		return i, i * i
+//	})
+func MapBy[T any, K comparable, V any](collection []T, fn func(T) (K, V)) map[K]V {
```
Comment (rickystewart) [inline pkg/util/slices.go:100]: Random q, does this name come from some place? I haven't heard of a function called `MapBy` before (contrast with `Filter` and `Map` which are very standard). I suppose I would prefer `MapFrom` as a name, as in "make a map from this collection"

**Review by rickystewart — APPROVED** [review]

File: pkg/util/slices.go (hunk 4 above)
Comment (chrisseto) [reply pkg/util/slices.go:100]: @JeffSwenson had suggested `AsMap`. `MapBy` just happens to be the thing I say to myself whenever I build a map from a slice, "map this by this key" which I realize is not the best statement 😅 

Updated to `MapFrom`!

File: pkg/util/slices_test.go (hunk 3 above)
Comment (chrisseto) [reply pkg/util/slices_test.go:131]: Done!

File: pkg/util/slices_test.go (hunk 2 above)
Comment (chrisseto) [reply pkg/util/slices_test.go:137]: Done!

**Comment by chrisseto** [comment]
TFTR!

bors r=rickystewart

I'll be out until the second week of October. If bors fails this and someone wants it, please feel free to merge it. Conversely, if someone really doesn't want this feel free to revert or r- it!

File: pkg/util/slices.go [hunk 5]
```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
...
+// Filter returns a new slice that only contains elements from collection that
+// satisfy predicate.
+//
+//	// Filter in place
+//	numbers = Filter(numbers, isEven)
+//	// Filter into a new slice
+//	odds := Filter(numbers, isEven)
+func Filter[T any](collection []T, predicate func(T) bool) []T {
+	i := 0
+	out := make([]T, len(collection))
```
Comment (pavelkalinnikov) [inline pkg/util/slices.go:64]: This allocates as much as the original slice, which can be wasteful if the `predicate` filters out most of the elements.

The alternative would be to create a nil slice, and let it grow/reallocate while we add things to it. This can be considered wasteful in other cases when `predicate` doesn't filter out much.

The "universal" strategy here is taking the `out` slice as a parameter:

```golang
func Filter[T any](collection []T, predicate func(T) bool, out []) []T {
	// if predicate(item) {
	// 	out = append(out, item)
	// }
}

// Easy, but appends to the nil slice, so may allocate multiple times.
even := Filter(numbers, isEven, nil)
// Zero allocations, reuses the original slice.
numbers = Filter(numbers, isEven, numbers[:0])
// Exactly one allocation if we know/guessed the number of elements 
even := Filter(numbers, isEven, make([]int, 0, 10))
```

File: pkg/util/slices.go [hunk 6]
```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
...
+	for j := range collection {
+		if predicate(collection[j]) {
+			out[i] = collection[j]
+			i++
+		}
+	}
+	return slices.Clip(out[:i])
```
Comment (pavelkalinnikov) [inline pkg/util/slices.go:71]: The `Clip` seems unnecessary here, just return `out[:i]` would be sufficient.

The implementation of `Clip` [returns](https://cs.opensource.google/go/x/exp/+/92128663:slices/slices.go;l=403-405) the same slice using a full slice expression which forces any appends to reallocate it. Usually this is used to protect elements beyond the slice length that are "owned" by someone else. Since the slice is allocated within this function, and is passed entirely to the caller, there is nothing to protect. Moreover, the caller could save an allocation if they want to append something to this slice after the call; but `Clip` forces a reallocation in this case.
